File: code/backend/webflux/v5/search/webflux_search_search_index/src/main/java/at/bumzack/foryouandyourfakewebshop/searchindexindex/SolrService.java
package at.bumzack.foryouandyourfakewebshop.searchindexindex;

import at.bumzack.common.solr.SolrReflectionThingi;
import at.bumzack.common.solr.SolrRequestBuilder;
import at.bumzack.common.solr.SolrResponse;
import at.bumzack.foryouandyourfakewebshop.searchindexindex.model.SearchDoc;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Mono;
import reactor.util.Logger;
import reactor.util.Loggers;

import java.util.stream.Collectors;

import static at.bumzack.common.solr.SolrExecutor.executeSolrGet;
import static java.util.Objects.nonNull;


@Component
public class SolrService {
    private static final Logger LOG = Loggers.getLogger(SolrService.class);

    @Value("${solr.host}")
    private String solrHost;

    @Value("${solr.port}")
    private Integer solrPort;

    public Mono<SolrResponse<SearchDoc>> searchByText(final String searchText, final int limit, final int offset, final ParameterizedTypeReference<SolrResponse<SearchDoc>> typeRef) {
        final var fieldNames = SolrReflectionThingi.getFieldNames(SearchDoc.class, String.class);

        final var q = fieldNames.stream()
                .map(name -> String.join(":", name, searchText))
                .collect(Collectors.joining(" OR "));

        LOG.info("Solr search query for String fields... i hope.  q = {}", q);

        final var builder = new SolrRequestBuilder()
                .setCoreName("searchindex")
                .setQuery(q)
                .setOffset(offset)
                .setLimit(limit);
        if (nonNull(solrHost)) {
            builder.setHost(solrHost);
        }

        if (nonNull(solrPort)) {
            builder.setPort(solrPort.toString());
        }

        final var url = builder
                .build();

        return executeSolrGet(typeRef, url);
    }
}

File: code/backend/csharp/SearchIndexService/SearchIndexService/Solr/SolrService.cs
namespace SearchIndexService.Solr;

public class SolrService
{
    private string _solrHost = "solr01.bumzack.at";

    private int _solrPort = 80;

    public async Task<SolrResponse<SearchDoc>> SearchByText(string searchText, int limit, int offset)
    {
        IEnumerable<string> fieldNames = new List<string>
        {
            "tconst",
            "titles",
            "actors",
            "directors",
            "genres",
            "characters"
        };

        IEnumerable<string> queries = fieldNames.Select(fn => { return string.Join(":", fn, searchText); });

        Console.WriteLine("queries  " + queries);
        string query = String.Join(" OR ", queries);
        Console.WriteLine("put together the query " + query);


        var builder = new SolrRequestBuilder()
            .SetCoreName("searchindex")
            .SetQuery(query)
            .SetOffset(offset)
            .SetLimit(limit);

        builder.SetHost(_solrHost);
        builder.SetPort(_solrPort);

        var url = builder
            .Build();

        Console.WriteLine("full Solr Request URL "+ url);
        var ex = new SolrRequestExecutor();

        return await ex.GetAsync(url);
    }
}